[Android] Xamarin caused by: android.runtime.JavaProxyThrowable: System.ObjectDisposedException: Cannot access a disposed object.

### Description

This error occurs when navigating between pages with complex-row listviews. Since no user code is involved it's not possible to catch and handle the exception.

### Steps to Reproduce

Intermittent.

### Expected Behavior

Navigating pages does not cause crashes.

### Actual Behavior

A crash occurs once in a while.

### Basic Information

- Version with issue: 2.5 280
- Last known good version: N/A
- IDE: VS2017 Community
- Platform Target Frameworks:
  - Android: 8.0
- Android Support Library Version: 26.1.0.1

### Reproduction
https://github.com/xamarin/Xamarin.Forms/files/2197967/disp2.zip

Exception happens here

![image](https://user-images.githubusercontent.com/5375137/42833562-a411dd4c-89b1-11e8-82e9-8072c0049144.png)

### Stack trace

```
Xamarin caused by: android.runtime.JavaProxyThrowable: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Android.Graphics.Bitmap'.
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <e8c0e16a54534fa885244f0ad837c79e>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeNonvirtualInt32Method (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <e8c0e16a54534fa885244f0ad837c79e>:0 
  at Android.Graphics.Bitmap.get_Height () [0x0000a] in <848bbd7c681a4975918c72f17d2f5144>:0 
  at Xamarin.Forms.Platform.Android.ButtonDrawable.Draw (Android.Graphics.Canvas canvas) [0x00021] in D:\agent_work\1\s\Xamarin.Forms.Platform.Android\Renderers\ButtonDrawable.cs:55 
  at Android.Graphics.Drawables.Drawable.n_Draw_Landroid_graphics_Canvas_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_canvas) [0x0000f] in <848bbd7c681a4975918c72f17d2f5144>:0 
  at (wrapper dynamic-method) System.Object:224a5ebe-c56f-4264-ad89-b784cc0da834 (intptr,intptr,intptr)
	at md5b60ffeb829f638581ab2bb9b1a7f4f3f.ButtonDrawable.n_draw(Native Method)
	at md5b60ffeb829f638581ab2bb9b1a7f4f3f.ButtonDrawable.draw(ButtonDrawable.java:50)
	at android.graphics.drawable.RippleDrawable.drawContent(RippleDrawable.java:836)
	at android.graphics.drawable.RippleDrawable.draw(RippleDrawable.java:692)
	at android.view.View.getDrawableRenderNode(View.java:18643)
	at android.view.View.drawBackground(View.java:18579)
	at android.view.View.draw(View.java:18367)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.draw(View.java:18379)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.widget.ListView.drawChild(ListView.java:3761)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.widget.AbsListView.dispatchDraw(AbsListView.java:3551)
	at android.widget.ListView.dispatchDraw(ListView.java:3744)
	at android.view.View.draw(View.java:18379)
	at android.widget.AbsListView.draw(AbsListView.java:7199)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.draw(View.java:18379)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.draw(View.java:18379)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.support.v4.widget.DrawerLayout.drawChild(DrawerLayout.java:1366)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.draw(View.java:18379)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.updateDisplayListIfDirty(View.java:17349)
	at android.view.View.draw(View.java:18138)
	at android.view.ViewGroup.drawChild(ViewGroup.java:3969)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3755)
	at android.view.View.draw(View.java:18379)
	at com.android.internal.policy.DecorView.draw(DecorView.java:940)
	at android.view.View.updateDisplayListIfDirty(View.java:17354)
	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:692)
	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:698)
	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:806)
	at android.view.ViewRootImpl.draw(ViewRootImpl.java:3136)
	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2932)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2520)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1519)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7113)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:927)
	at android.view.Choreographer.doCallbacks(Choreographer.java:702)
	at android.view.Choreographer.doFrame(Choreographer.java:638)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:913)
	at android.os.Handler.handleCallback(Handler.java:751)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6780)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1496)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1386)
```

Only happen on Android (tested on Samsung Galaxy S6 - Android 7.0), sometimes after 2 min, sometimes running for hours without crash.
I'm not using ListView, thought it was linked to OnPropertyChanged not running on UI thread so i tried :

```
#if __ANDROID__
            if (Android.OS.Looper.MyLooper() != null && Android.OS.Looper.MyLooper().Thread == Android.OS.Looper.MainLooper.Thread)
            {
                changed.Invoke(this, new PropertyChangedEventArgs(propertyName));
            }
            else
            {
                Device.BeginInvokeOnMainThread(() =>
                {
                    changed.Invoke(this, new PropertyChangedEventArgs(propertyName));
                });
            }
#endif
```


but even making it run on UI thread doesn't fix the issue.

```
Unhandled Exception:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.FormsTextView'.
  at (wrapper dynamic-method) System.Object:a3b623bc-af3e-44cb-ae90-81e9b2c34eb0 (intptr,intptr)
[ERROR] FATAL UNHANDLED EXCEPTION: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.FormsTextView'.
  at (wrapper dynamic-method) System.Object:a3b623bc-af3e-44cb-ae90-81e9b2c34eb0 (intptr,intptr)


05-28 14:06:36.211	Samsung SM-G920F	Error	23539	AndroidRuntime	Object name: 'Xamarin.Forms.Platform.Android.FormsTextView'.
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <e8c0e16a54534fa885244f0ad837c79e>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeNonvirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <e8c0e16a54534fa885244f0ad837c79e>:0 
  at Android.Widget.TextView.set_TextFormatted (Java.Lang.ICharSequence value) [0x00029] in <848bbd7c681a4975918c72f17d2f5144>:0 
  at Android.Widget.TextView.set_Text (System.String value) [0x0000d] in <848bbd7c681a4975918c72f17d2f5144>:0 
  at Xamarin.Forms.Platform.Android.LabelRenderer.UpdateText () [0x00098] in <aae4e9ce5a5b4d3eaec4edf21079f3fa>:0 
  at Xamarin.Forms.Platform.Android.LabelRenderer.OnElementPropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x000c5] in <aae4e9ce5a5b4d3eaec4edf21079f3fa>:0 
  at (wrapper delegate-invoke) <Module>:invoke_void_object_PropertyChangedEventArgs (object,System.ComponentModel.PropertyChangedEventArgs)
  at Xamarin.Forms.BindableObject.OnPropertyChanged (System.String propertyName) [0x00012] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.Element.OnPropertyChanged (System.String propertyName) [0x00000] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.Internals.SetValueFlags attributes, System.Boolean silent) [0x0010e] in <1aabc0fecdaa41b6a5c04760c77bb648>:0
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.Internals.SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes) [0x0015b] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.BindingExpression.ApplyCore (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property, System.Boolean fromTarget) [0x001f9] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.BindingExpression.Apply (System.Boolean fromTarget) [0x0003e] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.BindingExpression+BindingExpressionPart.<PropertyChanged>b__47_0 () [0x00000] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <848bbd7c681a4975918c72f17d2f5144>:0 
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <848bbd7c681a4975918c72f17d2f5144>:0 
  at (wrapper dynamic-method) System.Object:c9b101f5-2323-49dd-9e53-f56cac4b4160 (intptr,intptr)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:30)
	at android.os.Handler.handleCallback(Handler.java:751)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6682)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
```

--- a/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue2004.cs
+++ b/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue2004.cs
@@ -1,0 +1,301 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using Xamarin.Forms.CustomAttributes;
+using Xamarin.Forms.Internals;
+
+
+#if UITEST
+using Xamarin.UITest;
+using NUnit.Framework;
+using Xamarin.Forms.Core.UITests;
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Preserve(AllMembers = true)]
+	[Issue(IssueTracker.Github, 2004, "[Android] Xamarin caused by: android.runtime.JavaProxyThrowable: System.ObjectDisposedException: Cannot access a disposed object",
+		PlatformAffected.Android)]
+#if UITEST
+	[NUnit.Framework.Category(UITestCategories.LifeCycle)]
+#endif
+	public class Issue2004 : TestContentPage
+	{
+		static internal NavigationPage settingsPage = new NavigationPage(new SettingsView());
+		static internal NavigationPage addressesPage = new NavigationPage(new AddressListView());
+		static internal NavigationPage associationsPage = new NavigationPage(new ContentPage());
+		static MasterDetailPage RootPage;
+		protected override void Init()
+		{
+			MasterDetailPage testPage = new MasterDetailPage();
+			RootPage = testPage;
+			testPage.Master = new ContentPage
+			{
+				Title = "M",
+			};
+
+			testPage.Detail = new SettingsView();
+		}
+
+		protected override void OnAppearing()
+		{
+			base.OnAppearing();
+			Application.Current.MainPage = RootPage;
+		}
+
+		static void SetPage(Page page)
+		{
+			RootPage.Detail = page;
+		}
+
+		static async Task UI(int delay)
+		{
+			await Task.Delay(delay);
+		}
+
+		public static INavigation NavigationPage => RootPage.Detail.Navigation;
+
+		public static async Task DisposedBitmapTest()
+		{
+			SetPage(Issue2004.associationsPage);
+			await UI(999);
+			SetPage(Issue2004.addressesPage);
+			await UI(999);
+
+			SetPage(Issue2004.associationsPage);
+
+			await UI(999);
+			SetPage(Issue2004.addressesPage);
+			await UI(999);
+
+			await NavigationPage.PushAsync(new ContentPage());
+			await UI(999);
+			await NavigationPage.PopAsync();
+			await UI(999);
+
+			SetPage(Issue2004.associationsPage);
+			await UI(999);
+
+			SetPage(Issue2004.addressesPage);
+			await UI(999);
+			SetPage(new ContentPage() { Content = new Label() { Text = "Success" } });
+		}
+
+
+		[Preserve(AllMembers = true)]
+		public class AddressListItemView : Grid
+		{
+			public AddressListItemView()
+			{
+				this.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Auto });
+				this.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Auto });
+				this.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Auto });
+				this.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Auto });
+				this.ColumnDefinitions.Add(new ColumnDefinition() { Width = GridLength.Star });
+
+				this.Children.Add(new Button() { Text = "qwe", BackgroundColor = Color.Transparent }, 0, 0);
+				this.Children.Add(new Button() { Text = "qwe", BackgroundColor = Color.Transparent }, 1, 0);
+				this.Children.Add(new Button() { Text = "qwe", BackgroundColor = Color.Transparent }, 2, 0);
+				this.Children.Add(new Button() { Text = "qwe", BackgroundColor = Color.Transparent }, 3, 0);
+
+				this.Children.Add(new StackLayout()
+				{
+					Children =
+					{
+						new Label{ Text = "Address", LineBreakMode = LineBreakMode.TailTruncation},
+						new Label{ Text = "Owner", LineBreakMode = LineBreakMode.TailTruncation},
+						new Label{ Text = "ViolationCount"},
+					}
+
+				}, 4, 0);
+			}
+		}
+
+		[Preserve(AllMembers = true)]
+		public class AddressListView : ContentPage
+		{
+			public AddressListView()
+			{
+				ListView listView = new ListView() { RowHeight = 75 };
+
+				listView.SetBinding(ListView.ItemsSourceProperty, "UnitList");
+
+				listView.ItemTemplate = new DataTemplate(() =>
+				{
+					ViewCell cell = new ViewCell();
+					cell.View = new AddressListItemView();
+					return cell;
+				});
+
+				Content = new StackLayout()
+				{
+
+					Children =
+					{
+						new StackLayout()
+						{
+							Orientation = StackOrientation.Horizontal,
+							Padding = 4,
+							Children =
+							{
+								new StackLayout()
+								{
+									Children =
+									{
+										new Label()
+										{
+											Text = "SortText",
+											HorizontalOptions = LayoutOptions.Center
+										}
+									}
+								}
+							}
+						},
+						listView
+					}
+				};
+
+				BindingContext = this;
+			}
+
+			protected override void OnAppearing()
+			{
+				base.OnAppearing();
+				UnitList = null;
+				NotifyPropertyChanged(() => UnitList);
+				SelectedAddress = null;
+				LoadAddresses();
+			}
+
+			string _selectedAddress;
+			public string SelectedAddress
+			{
+				get => _selectedAddress;
+				set
+				{
+					if (_selectedAddress != value)
+					{
+						_selectedAddress = value;
+						NotifyPropertyChanged(() => SelectedAddress);
+						if (SelectedAddress != null)
+						{
+							LoadUnitsByAddress(_selectedAddress);
+							NotifyPropertyChanged(() => UnitList);
+						}
+					}
+				}
+			}
+
+			List<string> _streeAddresses;
+			private List<string> _unitList;
+
+			public List<string> StreetAddresses
+			{
+				get { return _streeAddresses; }
+				set
+				{
+					_streeAddresses = value;
+					NotifyPropertyChanged();
+				}
+			}
+
+			public void LoadAddresses()
+			{
+				StreetAddresses = Enumerable.Range(1, 10).Select(x => x.ToString()).ToList();
+				SelectedAddress = StreetAddresses.First();
+			}
+
+			public void LoadUnitsByAddress(string address)
+			{
+				if (string.IsNullOrEmpty(address))
+				{
+					UnitList?.Clear();
+					return;
+				}
+				UnitList = Enumerable.Range(1, 10).Select(x => x.ToString()).ToList();
+			}
+
+			public List<string> UnitList
+			{
+				get { return _unitList; }
+				set { _unitList = value; }
+			}
+
+			public virtual void NotifyPropertyChanged([CallerMemberName] string propertyName = null)
+			{
+				OnPropertyChanged(propertyName);
+			}
+
+			protected virtual void NotifyPropertyChanged<T>(Expression<Func<T>> propertyExpression)
+			{
+				string propertyName = GetPropertyName(propertyExpression);
+				OnPropertyChanged(propertyName);
+			}
+
+			private string GetPropertyName<T>(Expression<Func<T>> propertyExpression)
+			{
+				if (propertyExpression == null)
+				{
+					throw new ArgumentNullException("propertyExpression");
+				}
+
+				if (propertyExpression.Body.NodeType != ExpressionType.MemberAccess)
+				{
+					throw new ArgumentException("Should be a member access lambda expression", "propertyExpression");
+				}
+
+				var memberExpression = (MemberExpression)propertyExpression.Body;
+				return memberExpression.Member.Name;
+			}
+		}
+
+
+		[Preserve(AllMembers = true)]
+		public class SettingsView : ContentPage
+		{
+			public Command AutoTest => new Command(async () =>
+			{
+				await Issue2004.DisposedBitmapTest();
+			});
+
+
+			protected async override void OnAppearing()
+			{
+				base.OnAppearing();
+				await Task.Delay(1000);
+				AutoTest.Execute(null);
+
+			}
+			public SettingsView()
+			{
+				BindingContext = this;
+				Content = new ScrollView()
+				{
+					Content = new StackLayout()
+					{
+						Children =
+						{
+							new Label()
+							{
+								Text = "Auto Test",
+								HorizontalOptions = LayoutOptions.Start
+							}
+						}
+					}
+				};
+			}
+		}
+
+#if UITEST
+		[Test]
+		public void NoCrashFromDisposedBitmapWhenSwitchingPages()
+		{
+			RunningApp.WaitForElement("Success");
+		}
+#endif
+	}
+}

--- a/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems
+++ b/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems
@@ -9,6 +9,7 @@
     <Import_RootNamespace>Xamarin.Forms.Controls.Issues</Import_RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)Issue2004.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Issue3333.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Issue2338.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Bugzilla60045.xaml.cs">

--- a/Xamarin.Forms.Controls/App.cs
+++ b/Xamarin.Forms.Controls/App.cs
@@ -46,6 +46,9 @@ namespace Xamarin.Forms.Controls
 
 			//// Uncomment to verify that there is no crash when switching MainPage from MDP inside NavPage
 			//SetMainPage(new Bugzilla45702());
+
+			//// Uncomment to verify that there is no crash when rapidly switching pages that contain lots of buttons
+			//SetMainPage(new Issues.Issue2004());
 		}
 
 		public Page CreateDefaultMainPage()

--- a/Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs
@@ -162,7 +162,7 @@ namespace Xamarin.Forms.Platform.Android.AppCompat
 			if (Element == null || Control == null)
 				return;
 
-			_backgroundTracker?.UpdateBackgroundColor();
+			_backgroundTracker?.UpdateDrawable();
 		}
 
 		void UpdateAll()

--- a/Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs
@@ -173,16 +173,7 @@ namespace Xamarin.Forms.Platform.Android.AppCompat
 			UpdateTextColor();
 			UpdateEnabled();
 			UpdateBackgroundColor();
-			UpdateDrawable();
 			UpdatePadding();
-		}
-
-		void UpdateDrawable()
-		{
-			if (Element == null || Control == null)
-				return;
-
-			_backgroundTracker?.UpdateDrawable();
 		}
 
 		void UpdateBitmap()

--- a/Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs
@@ -95,6 +95,7 @@ namespace Xamarin.Forms.Platform.Android.AppCompat
 					_textColorSwitcher = null;
 				}
 				_backgroundTracker?.Dispose();
+				_backgroundTracker = null;
 			}
 
 			base.Dispose(disposing);

--- a/Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/AppCompat/NavigationPageRenderer.cs
@@ -1072,12 +1072,15 @@ namespace Xamarin.Forms.Platform.Android.AppCompat
 				if (shouldUpdateToolbar)
 					UpdateToolbar();
 
-				FragmentTransaction fragmentTransaction = fragmentManager.BeginTransactionEx();
+				if (fragmentsToRemove.Count > 0)
+				{
+					FragmentTransaction fragmentTransaction = fragmentManager.BeginTransactionEx();
 
-				foreach (Fragment frag in fragmentsToRemove)
-					fragmentTransaction.RemoveEx(frag);
+					foreach (Fragment frag in fragmentsToRemove)
+						fragmentTransaction.RemoveEx(frag);
 
-				fragmentTransaction.CommitAllowingStateLossEx();
+					fragmentTransaction.CommitAllowingStateLossEx();
+				}
 
 				return false;
 			});

--- a/Xamarin.Forms.Platform.Android/ButtonBackgroundTracker.cs
+++ b/Xamarin.Forms.Platform.Android/ButtonBackgroundTracker.cs
@@ -120,11 +120,6 @@ namespace Xamarin.Forms.Platform.Android
 			}
 		}
 
-		public void UpdateBackgroundColor()
-		{
-			UpdateDrawable();
-		}
-
 		public void Dispose()
 		{
 			Dispose(true);

--- a/Xamarin.Forms.Platform.Android/FastRenderers/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/FastRenderers/ButtonRenderer.cs
@@ -198,6 +198,7 @@ namespace Xamarin.Forms.Platform.Android.FastRenderers
 				_tracker?.Dispose();
 
 				_backgroundTracker?.Dispose();
+				_backgroundTracker = null;
 
 				if (Element != null)
 				{

--- a/Xamarin.Forms.Platform.Android/FastRenderers/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/FastRenderers/ButtonRenderer.cs
@@ -45,7 +45,7 @@ namespace Xamarin.Forms.Platform.Android.FastRenderers
 		{
 			_automationPropertiesProvider = new AutomationPropertiesProvider(this);
 			_effectControlProvider = new EffectControlProvider(this);
-			
+
 			Initialize();
 		}
 
@@ -244,7 +244,6 @@ namespace Xamarin.Forms.Platform.Android.FastRenderers
 				UpdateIsEnabled();
 				UpdateInputTransparent();
 				UpdateBackgroundColor();
-				UpdateDrawable();
 				UpdatePadding();
 
 				ElevationHelper.SetElevation(this, e.NewElement);
@@ -321,7 +320,7 @@ namespace Xamarin.Forms.Platform.Android.FastRenderers
 
 		void UpdateBackgroundColor()
 		{
-			_backgroundTracker?.UpdateBackgroundColor();
+			_backgroundTracker?.UpdateDrawable();
 		}
 
 		internal void OnNativeFocusChanged(bool hasFocus)
@@ -484,7 +483,7 @@ namespace Xamarin.Forms.Platform.Android.FastRenderers
 			_textColorSwitcher.Value.UpdateTextColor(this, Button.TextColor);
 		}
 
-		void UpdatePadding ()
+		void UpdatePadding()
 		{
 			SetPadding(
 				(int)(Context.ToPixels(Button.Padding.Left) + _paddingDeltaPix.Left),
@@ -494,15 +493,10 @@ namespace Xamarin.Forms.Platform.Android.FastRenderers
 			);
 		}
 
-		void UpdateContentEdge (Thickness? delta = null)
+		void UpdateContentEdge(Thickness? delta = null)
 		{
-			_paddingDeltaPix = delta ?? new Thickness ();
-			UpdatePadding ();
-		}
-
-		void UpdateDrawable()
-		{
-			_backgroundTracker?.UpdateDrawable();
+			_paddingDeltaPix = delta ?? new Thickness();
+			UpdatePadding();
 		}
 
 	}

--- a/Xamarin.Forms.Platform.Android/Renderers/ButtonDrawable.cs
+++ b/Xamarin.Forms.Platform.Android/Renderers/ButtonDrawable.cs
@@ -63,15 +63,25 @@ namespace Xamarin.Forms.Platform.Android
 			if (width <= 0 || height <= 0)
 				return;
 
-			if (_normalBitmap == null || _normalBitmap.Height != height || _normalBitmap.Width != width)
-			{
+			if (_normalBitmap == null ||
+				_normalBitmap?.IsDisposed() == true ||
+				_pressedBitmap?.IsDisposed() == true ||
+				_normalBitmap.Height != height ||
+				_normalBitmap.Width != width)
 				Reset();
 
-				_normalBitmap = CreateBitmap(false, width, height);
-				_pressedBitmap = CreateBitmap(true, width, height);
+			Bitmap bitmap = null;
+			if (GetState().Contains(global::Android.Resource.Attribute.StatePressed))
+			{
+				_pressedBitmap = _pressedBitmap ?? CreateBitmap(true, width, height);
+				bitmap = _pressedBitmap;
+			}
+			else
+			{
+				_normalBitmap = _normalBitmap ?? CreateBitmap(false, width, height);
+				bitmap = _normalBitmap;
 			}
 
-			Bitmap bitmap = GetState().Contains(global::Android.Resource.Attribute.StatePressed) ? _pressedBitmap : _normalBitmap;
 			canvas.DrawBitmap(bitmap, 0, 0, new Paint());
 		}
 
@@ -95,15 +105,21 @@ namespace Xamarin.Forms.Platform.Android
 		{
 			if (_normalBitmap != null)
 			{
-				_normalBitmap.Recycle();
-				_normalBitmap.Dispose();
+				if (!_normalBitmap.IsDisposed())
+				{
+					_normalBitmap.Recycle();
+					_normalBitmap.Dispose();
+				}
 				_normalBitmap = null;
 			}
 
 			if (_pressedBitmap != null)
 			{
-				_pressedBitmap.Recycle();
-				_pressedBitmap.Dispose();
+				if (!_pressedBitmap.IsDisposed())
+				{
+					_pressedBitmap.Recycle();
+					_pressedBitmap.Dispose();
+				}
 				_pressedBitmap = null;
 			}
 		}

--- a/Xamarin.Forms.Platform.Android/Renderers/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/Renderers/ButtonRenderer.cs
@@ -147,7 +147,7 @@ namespace Xamarin.Forms.Platform.Android
 			if (Element == null || Control == null)
 				return;
 
-			_backgroundTracker?.UpdateBackgroundColor();
+			_backgroundTracker?.UpdateDrawable();
 		}
 
 		void UpdateAll()
@@ -157,7 +157,7 @@ namespace Xamarin.Forms.Platform.Android
 			UpdateBitmap();
 			UpdateTextColor();
 			UpdateEnabled();
-			UpdateDrawable();
+			UpdateBackgroundColor();
 			UpdatePadding();
 		}
 
@@ -217,11 +217,6 @@ namespace Xamarin.Forms.Platform.Android
 			image?.Dispose();
 		}
 
-		void UpdateDrawable()
-		{
-			_backgroundTracker.UpdateDrawable();
-		}
-
 		void UpdateEnabled()
 		{
 			Control.Enabled = Element.IsEnabled;
@@ -270,7 +265,7 @@ namespace Xamarin.Forms.Platform.Android
 
 		void UpdatePadding()
 		{
-			Control?.SetPadding (
+			Control?.SetPadding(
 				(int)(Context.ToPixels(Element.Padding.Left) + _paddingDeltaPix.Left),
 				(int)(Context.ToPixels(Element.Padding.Top) + _paddingDeltaPix.Top),
 				(int)(Context.ToPixels(Element.Padding.Right) + _paddingDeltaPix.Right),
@@ -278,9 +273,9 @@ namespace Xamarin.Forms.Platform.Android
 			);
 		}
 
-		void UpdateContentEdge (Thickness? delta = null)
+		void UpdateContentEdge(Thickness? delta = null)
 		{
-			_paddingDeltaPix = delta ?? new Thickness ();
+			_paddingDeltaPix = delta ?? new Thickness();
 			UpdatePadding();
 		}
 

--- a/Xamarin.Forms.Platform.Android/Renderers/ButtonRenderer.cs
+++ b/Xamarin.Forms.Platform.Android/Renderers/ButtonRenderer.cs
@@ -82,6 +82,7 @@ namespace Xamarin.Forms.Platform.Android
 			if (disposing)
 			{
 				_backgroundTracker?.Dispose();
+				_backgroundTracker = null;
 			}
 
 			base.Dispose(disposing);

